lib/BackgroundJob/SendScheduledMessages.php

```diff
@@ -0,0 +1,214 @@
+use Psr\Log\LoggerInterface;
+
+/**
+ * Class SendScheduledMessages
+ *
+ * @package OCA\Talk\BackgroundJob
+ */
+class SendScheduledMessages extends TimedJob {
+	private readonly IL10N $l10n;
+	public function __construct(
+		private ScheduledMessageService $scheduledMessageService,
+		private ParticipantService $participantService,
+		private Manager $manager,
+		private ChatManager $chatManager,
+		private MessageParser $messageParser,
+		private ThreadService $threadService,
+		private IFactory $l10nFactory,
+		private LoggerInterface $logger,
+		ITimeFactory $time,
+	) {
+		// Every minute
+		$this->setInterval(60);
+		$this->l10n = $this->l10nFactory->get(Application::APP_ID);
+		parent::__construct($time);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	#[\Override]
+	protected function run($argument): void {
+		$time = $this->time->getDateTime('-1 second');
+		$messages = $this->scheduledMessageService->getDue($time);
+		if (empty($messages)) {
+			$this->logger->debug('No messages found');
+			return;
+		}
+
+		/** @var list<Room> $rooms */
+		$rooms = [];
+		$toDelete = [];
+		foreach ($messages as $message) {
+			if (!isset($rooms[$message->getRoomId()])) {
+				try {
+					$rooms[$message->getRoomId()] = $this->manager->getRoomById($message->getRoomId());
+				} catch (RoomNotFoundException) {
+					$this->logger->error('Room not found: ' . $message->getRoomId());
+					continue;
+				}
+			}
+
+			$room = $rooms[$message->getRoomId()];
+			try {
+				$participant = $this->participantService->getParticipantByActor($room, $message->getActorType(), $message->getActorId());
+			} catch (ParticipantNotFoundException $e) {
+				$this->logger->error('Participant not found', ['exception' => $e]);
+				$this->scheduledMessageService->deleteMessage(
+					$room,
+					(string)$message->getId(),
+					$message->getActorType(),
+					$message->getActorId()
+				);
+				continue;
+			}
+
+			if ($room->isFederatedConversation()) {
+				$this->logger->warning('Cannot send scheduled messages to federated conversations');
+				$this->scheduledMessageService->deleteMessage(
+					$room,
+					(string)$message->getId(),
+					$participant->getAttendee()->getActorType(),
+					$participant->getAttendee()->getActorId()
+				);
+				$this->participantService->setHasScheduledMessages($participant, Participant::ERROR_SCHEDULED_MESSAGE);
+				continue;
+			}
+
+			if ($room->getReadOnly() === Room::READ_ONLY) {
+				$this->logger->warning('Cannot send scheduled messages to read only room');
+				$this->scheduledMessageService->markAsFailed($message);
+				$this->participantService->setHasScheduledMessages($participant, Participant::ERROR_SCHEDULED_MESSAGE);
+				continue;
+			}
+
+			if ($room->getLobbyState() !== Webinary::LOBBY_NONE && ($participant->getPermissions() & Attendee::PERMISSIONS_LOBBY_IGNORE) === 0) {
+				$this->logger->warning('User has no chat permissions for room ' . $message->getRoomId() . ', could not send scheduled message');
```

Comment:
should log which user it was, since it's in a background job we don't have it otherwise.